expand/expand.1
.Dd December 2008
.Dt EXPAND 1

.Sh NAME

.Nm expand
.Nd convert tabs to spaces

.Sh SYNOPSIS

.Nm expand
.Ar [file...]

.Sh DESCRIPTION

The expand utility shall write files or the standard input to the standard output with <tab> characters replaced with one or more <space> characters needed to pad to the next tab stop. Any <backspace> characters shall be copied to the output and cause the column position count for tab stop calculations to be decremented; the column position count shall not be decremented below zero.

.Sh OPTIONS
The expand utility shall conform to XBD Utility Syntax Guidelines.

The following option shall be supported:

-t  tablist
Specify the tab stops. The application shall ensure that the argument tablist consists of either a single positive decimal integer or a list of tabstops. If a single number is given, tabs shall be set that number of column positions apart instead of the default 8.
If a list of tabstops is given, the application shall ensure that it consists of a list of two or more positive decimal integers, separated by <blank> or <comma> characters, in ascending order. The <tab> characters shall be set at those specific column positions. Each tab stop N shall be an integer value greater than zero, and the list is in strictly ascending order. This is taken to mean that, from the start of a line of output, tabbing to position N shall cause the next character output to be in the (N+1)th column position on that line.

In the event of expand having to process a <tab> at a position beyond the last of those specified in a multiple tab-stop list, the <tab> shall be replaced by a single <space> in the output.

.Sh OPERANDS
The following operand shall be supported:

file
The pathname of a text file to be used as input.
.Sh STDIN
See the INPUT FILES section.

.Sh INPUT FILES
Input files shall be text files.

.Sh ENVIRONMENT VARIABLES
The following environment variables shall affect the execution of expand:

LANG
Provide a default value for the internationalization variables that are unset or null. (See XBD Internationalization Variables for the precedence of internationalization variables used to determine the values of locale categories.)
LC_ALL
If set to a non-empty string value, override the values of all the other internationalization variables.
LC_CTYPE
Determine the locale for the interpretation of sequences of bytes of text data as characters (for example, single-byte as opposed to multi-byte characters in arguments and input files), the processing of <tab> and <space> characters, and for the determination of the width in column positions each character would occupy on an output device.
LC_MESSAGES
Determine the locale that should be used to affect the format and contents of diagnostic messages written to standard error.
NLSPATH
[XSI] [Option Start] Determine the location of message catalogs for the processing of LC_MESSAGES. [Option End]
.Sh ASYNCHRONOUS EVENTS
Default.

.Sh STDOUT
The standard output shall be equivalent to the input files with <tab> characters converted into the appropriate number of <space> characters.

.Sh STDERR
The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES
None.

.Sh EXTENDED DESCRIPTION
None.

.Sh EXIT STATUS
The following exit values shall be returned:

 0
Successful completion
>0
An error occurred.
.Sh CONSEQUENCES OF ERRORS
The expand utility shall terminate with an error message and non-zero exit status upon encountering difficulties accessing one of the file operands.

The following sections are informative.
.Sh APPLICATION USAGE
None.

.Sh EXAMPLES
None.

.Sh RATIONALE
The expand utility is useful for preprocessing text files (before sorting, looking at specific columns, and so on) that contain <tab> characters.

See XBD Column Position.

The tablist option-argument consists of integers in ascending order. Utility Syntax Guideline 8 mandates that expand shall accept the integers (within the single argument) separated using either <comma> or <blank> characters.

Earlier versions of this standard allowed the following form in the SYNOPSIS:

expand [-tabstop][-tab1,tab2,...,tabn][file ...]
This form is no longer specified by POSIX.1-2017 but may be present in some implementations.

.Sh FUTURE DIRECTIONS
None.

.Sh SEE ALSO
tabs, unexpand

XBD Column Position, Environment Variables, Utility Syntax Guidelines



NEWS

Version 2019.02.10

	- Building with C++11 as minimum compiler and target.
	- Refresh almost all utilities, building them as C++
	- New all-in-one-makefile autotools build system.
	- Vastly improved testing framework.
	- Import man pages from Open Group (permitted per agreement).

	- New utilities: cmp, diff, expand, expr, ipcs, link, sleep,
	  sort, tput, tabs, tsort, stty, uname, uniq, unlink

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



expand/expand.cc
/*
 * Copyright 2019 Bloq Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include "posixutils-config.h"
#include <vector>
#include <string>
#include <regex>
#include <argp.h>
#include <libpu.h>

using namespace std;

static const char doc[] =
N_("expand - convert tabs to spaces");

static struct argp_option options[] = {
	{ "tabs", 't', "tablist", 0,
	  N_("Specify the tab stops. A single positive integer, or comma/blank-separated positive integers.") },
	{ }
};
DECL_ARGP(file_args_doc)

class ExpandApp : public CmdlineApp {
private:
	unsigned int		repeated_tab;
	unsigned int		last_tab;
	Bitmap			tab_map;

public:
	ExpandApp() : CmdlineApp(true), repeated_tab(8) { }

	int run() { return run_arg_files(); }
	int arg_file(StdioFile& f);

	bool set_tablist(const std::string& liststr);

private:
	void outspace(unsigned int& column);
	void advance_rtab(unsigned int& column);
	void advance_tablist(unsigned int& column);
};

static regex rxInteger("^(\\d+)$");

bool ExpandApp::set_tablist(const std::string& liststr)
{
	vector<string> sv;

	// branch: tab "list" is single positive integer
	if (regex_match(liststr, rxInteger)) {
		int tmp = atoi(liststr.c_str());
		if (tmp < 1)
			return false;
		last_tab = repeated_tab = tmp;
		return true;
	}
	
	// split based on comma, if present, blanks otherwise
	if (liststr.find(',')) {
		strsplit(liststr, ',', sv);
	} else
		strsplit(liststr, sv);

	// require at least two tablist entries
	if (sv.size() < 2)
		return false;

	// validate each tablist entry
	int last = 0;
	for (auto it = sv.begin(); it != sv.end(); it++) {
		// must be positive, consecutive integer
		int cur = atoi((*it).c_str());
		if ((cur < 1) || (cur <= last))
			return false;

		tab_map.set(cur);
		last_tab = cur;

		last = cur;
	}

	// we have a tablist; disable repeated-tab feature
	repeated_tab = 0;

	return true;
}

void ExpandApp::outspace(unsigned int& column)
{
	putchar(' ');
	column++;
}

void ExpandApp::advance_rtab(unsigned int& column)
{
	while ((column % repeated_tab) != 0)
		outspace(column);
	outspace(column);
}

void ExpandApp::advance_tablist(unsigned int& column)
{
	if (column >= last_tab) {
		outspace(column);
		return;
	}

	while (column < last_tab) {
		outspace(column);
		if (tab_map.test(column))
			break;
	}
	outspace(column);
}

// Core data processing algorithm and pipeline
int ExpandApp::arg_file(StdioFile& f)
{
	unsigned int column = 1;
	while (!f.eof()) {
		int ch = f.getc();
		if (ch == EOF)			// end of file, or error
			break;

		if (ch == '\b') {		// backspace
			putchar(ch);
			if (column > 1)
				column--;
		} else if (ch == '\r' || ch == '\n') { // newline
			putchar(ch);
			column = 1;
		} else if (ch != '\t') {	// all other non-tab chars
			putchar(ch);
			column++;
		} else if (repeated_tab)	// expand by repeated tab
			advance_rtab(column);
		else				// expand by tablist
			advance_tablist(column);
	}

	// return error at toplevel, upon any I/O error here
	return (f.err() ? 1 : 0);
}

static ExpandApp app;

static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	PU_OPT_BEGIN
	PU_APP_CALL_ARG('t', set_tablist)
	PU_APP_PUSH_ARG
	PU_OPT_DEFAULT
	PU_OPT_END
}

int main (int argc, char *argv[])
{
	return app.init_and_run(&argp, argc, argv);
}

